Adding support for conditional hasOne / hasMany mappings

AdamEisfeld opened the pull request:
This PR addresses the issue presented here: https://github.com/lucasmedeirosleite/EasyMapping/issues/141

By exposing the created EKRelationshipMapping object in the return value of the hasOne / hasMany methods on EKObjectMapping, adding a new "condition" block to EKRelationshipMapping that provides the current representation object and accepts a BOOL return value, and checking a given EKRelationshipMapping's condition block return value (if it has one) before using it, we can conditionally map classes to keys.

Note that the benefits of doing things this way is there is no required changes by existing clients if they update their EasyMapping implementation. Users can simply opt into this by providing condition blocks on their relationships via the returned EKRelationshipMapping objects if they want to.

In a later PR, convenience initializers for this kind of mapping can be added, but for the time being conditional mapping would now be possible doing something like this (in the objectMapping class method, for example):

```objc
+ (EKObjectMapping *)objectMapping {
    
    return [EKObjectMapping mappingForClass:self withBlock:^(EKObjectMapping *mapping) {
        
        // Conditionally map the "data" keypath to a Foo class, if the "dataType" is equal to "foo"
        EKRelationshipMapping *relationshipFoo = [mapping hasOne:[Foo class] forKeyPath:@"data"];
        relationship.condition = ^BOOL(id representation) {
            NSDictionary *dictionary = representation;
            if ([dictionary isKindOfClass:[NSDictionary class]]) {
                NSString *type = dictionary[@"dataType"];
                if ([type isKindOfClass:[NSString class]]) {
                    return [type isEqualToString:@"foo"];
                }
            }
            return NO;
        };
        
        // Conditionally map the "data" keypath to a Bar class, if the "dataType" is equal to "bar"
        EKRelationshipMapping *relationshipBar = [mapping hasOne:[Bar class] forKeyPath:@"data"];
        relationshipBar.condition = ^BOOL(id representation) {
            NSDictionary *dictionary = representation;
            if ([dictionary isKindOfClass:[NSDictionary class]]) {
                NSString *type = dictionary[@"dataType"];
                if ([type isKindOfClass:[NSString class]]) {
                    return [type isEqualToString:@"bar"];
                }
            }
            return NO;
        };
        
    }];
    
}
```

If approved, then a future PR can be made to provide simpler mechanisms for performing the same logic, eg;

```objc
+ (EKObjectMapping *)objectMapping {
    
    return [EKObjectMapping mappingForClass:self withBlock:^(EKObjectMapping *mapping) {
        
        // Conditionally map the "data" keypath to a Foo class, if the "dataType" is equal to "foo"
        EKRelationshipMapping *relationshipFoo = [mapping hasOne:[Foo class] forKeyPath:@"data"];
        relationshipFoo.condition = [EKRelationshipConditionHelper conditionIfValueForKey: @"dataType" isEqualToString: @"foo"];
        
        // Conditionally map the "data" keypath to a Bar class, if the "dataType" is equal to "bar"
        EKRelationshipMapping *relationshipBar = [mapping hasOne:[Foo class] forKeyPath:@"data"];
        relationshipBar.condition = [EKRelationshipConditionHelper conditionIfValueForKey: @"dataType" isEqualToString: @"bar"];
        
    }];
    
}
```

DenHeadless commented:
Hey, @AdamEisfeld!

Thanks for PR, this feature definitely seems interesting and i've seen real world examples of such API in my own projects. I think we want to include it in the next release. Here are however some comments you might want to adrress:

1. HasMany mapping handling does not seem to be correct. It assumes that only one mapping is used for entire relationship, which is not exactly true.

In your example in #142 you have many vehicles, and first mapping is used for car, and second - for plane. It seems like we need to somehow enumerate available hasManyMappings for keyPath we are now mapping, and select one that is needed for current object. 

2. Can i ask you to add some unit tests to this feature? We need to know, that it's working reliably to include it in the framework.

AdamEisfeld commented:
Hi @DenHeadless ,

1. You may be correct about the hasMany mapping, I'll look into that. Admittedly I don't have an entire grasp of how everything works under the hood with Easy Mapping yet.

2. Will do, I'll add these after the changes have been signed off in the PR (before the PR is merged of course).

Regarding the "next release", is there some sort of a schedule for when that will be? Or is it just based on whoever has time to create a new release / if there are enough new features to warrant one?

Thanks,
- Adam

DenHeadless commented:
No, no release schedule, when new feature comes along, we simply gather all we have on issue tracker and make a release. 

Next release could be for Xcode 8, which should be released in a week or so, we wanted to remove some deprecated API and drop iOS 7, however it's not pressing at all and can wait. Unlike big Swift 3 migration, EasyMapping Objective-C API does not need any adjustment =)

AdamEisfeld commented:
I just realized this approach wont work because the hasOne / hasMany relationships are stored in dictionaries instead of arrays where the key is the keypath being mapped, so any subsequent mappings using the same key path will just overwrite the previous one.

I am looking into an alternate solution now and will update the PR accordingly.

AdamEisfeld commented:
@DenHeadless I have updated the implementation accordingly:

1. hasOneMappings / hasManyMappings are now NSMutableArrays instead of NSDictionaries. This allows for multiple EKRelationshipMapping objects to point to the same key path, instead of overwriting any subsequent EKRelationshipMapping objects with the same key path. 

2. I added a check to EKSerializer to check if the class of the object being deserialized matches the class stored in the EKRelationshipMapping object. This is necessary to stop the optional relationships from attempting to deserialize.

I have tested the optional hasOne flow in my own app and it seems to work great. I haven't had a chance to test the hasMany flow yet but looking at the code I think it should work, but I will verify this next.

Pending your approval of these changes and my verification that the hasMany flow works, I will add unit tests to cover the optional mappings.

Let me know your thoughts, 
- Adam

AdamEisfeld commented:
Ah, it looks like I'll need to update the test cases to use the new arrays instead of dictionaries.

DenHeadless commented:
Hey, @AdamEisfeld, sorry for a lack of response, Swift 3 migration is taking all my free time right now.

Can i ask you to give me push access to your fork via Github "Allow edits from maintainers" feature?
http://d.pr/i/gAF5

I've made some changes locally to fix unit tests, and wanted to push them to this PR, and then proceed with discussion of this feature.

armandzwan commented:
> 1. HasMany mapping handling does not seem to be correct. It assumes that only one mapping is used for entire relationship, which is not exactly true. In your example in #142 you have many vehicles, and first mapping is used for car, and second - for plane. It seems like we need to somehow enumerate available hasManyMappings for keyPath we are now mapping, and select one that is needed for current object.

Am I missing something, or was this option never build? I want to serialize ToMany objects with different mappings, but it seems like I can only apply one mapping to the whole Set.

DenTelezhkin commented:
The best way to verify expected behavior is provide a unit test(s) for your use case, that either passes or breaks in unexpected way, so that we could fix it. I don’t remember exactly all applications of this feature unfortunately.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

armandzwan commented:
I've got it working of I send the Relationship to `serializeCollection` like so:
`+(NSArray *)serializeCollection:(NSArray *)collection withRelationship:(EKRelationshipMapping *)relationship fromContext:(NSManagedObjectContext *)context`
And then use `(EKManagedObjectMapping *)[relationship mappingForObject:object]` in the for loop. I will fork and apply those changes for you to see the changes. It will be a breaking change for those who use a custom `EKSerializationResolvingBlock` because now it is called per object in the loop, not just the whole set.

DenTelezhkin commented:
Please do, let's see what it will break in our test suite to determine the path forward.

armandzwan commented:
Can you have a look at:
https://github.com/armandzwan/EasyMapping/tree/serializationResolver

DenTelezhkin commented:
I agree with your conclusions, custom serialization resolvers need some changes to work properly with collections. But also, since this framework is currently not under active development, backwards compatibility is really important.

I've pulled changes from your fork into master, but kept original EKSerializer methods around with deprecation warnings. If anyone is using object serialization, which underneath is calling collection serialization, they will not even see deprecation warnings, stuff will continue to work. But for someone who used `EKSerializer.serializeCollection:withMapping:` method deprecation warning will show up, and user can choose to silence it or rework mappings to call  `EKSerializer.serializeCollection:withRelationship:` method. I've also made sure test suite and commit with those changes include example how this can be done.

@armandzwan It would be nice if you could try out changes from master branch(commit a08189243cf64f794267a27c05c6ebff7a747f1b), and if everything's fine - i'll make a new release.
